Update v2 developers guide (#351)
NCI-GDC/gdc-docs: docs/Data_Portal/Users_Guide/Developers_Guide.md

# GDC Analysis Tool Software Development Kit (SDK)

This guide will detail the process of developing applications for the GDC Data Portal 2.0. It describes the structure of
the GDC Data Portal, how to use the GDC Data Portal API, and how to develop applications for the GDC Data Portal.

The GDC Data Portal is a repository and computational platform for cancer researchers who need to understand cancer, its
clinical progression, and response to therapy. The GDC Data Portal supports the development of applications that allow
for analysis, visualization, and refinement of cohorts.

The GDC Data Portal is built on top of the [GDC API](https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/),
which provides access to the GDC data. The GDC Data Portal provides an Analysis Tool Framework (ATF) for developing
applications that can be used to analyze, visualize, and download data from the GDC.

The GDC Data Portal is built with the [React](https://reactjs.org/) framework and 

> the [Redux](https://redux.js.org/) library for state management. The GDC Data Portal uses [NextJS](https://nextjs.org/) as it's application framework which 

Reviewer: ```suggestion
the [Redux](https://redux.js.org/) library for state management. The GDC Data Portal uses [NextJS](https://nextjs.org/) as it's the application framework which 
```

> the [Redux](https://redux.js.org/) library for state management. The GDC Data Portal uses [NextJS](https://nextjs.org/) as it's application framework which 

Reviewer: ```suggestion
the [Redux](https://redux.js.org/) library for state management. The GDC Data Portal uses [NextJS](https://nextjs.org/) as its application framework which 
```

provides server-side rendering of React components. [Mantine.dev](https://mantine.dev/) is the component library, and

styling is through [TailwindCSS](https://tailwindcss.com/). The GDC Data Portal is built on top of the GDC API, which provides access to 
the GDC data.

![This image details the architecture of the GDC Data Portal. 
It shows the interaction between the GDC Data API, the core
module and the user interface.](./images/developers_guide/V2_architecture.png "Architecture of the GDC Data Portal")

*Architecture of the GDC Data Portal*

## Overview of an Application

Applications are React higher-order components (HOC) that are rendered in

> the [Analysis Center](https://portal.gdc.cancer.gov/analysis_page?app=). The GDC Data Portal major functions such as
> Projects, Repository,  and Protein Paint are all applications. Each application handles a specific task such as analysis or

Reviewer: ```suggestion
the [Analysis Center](https://portal.gdc.cancer.gov/analysis_page?app=). The GDC Data Portal's major functions such as
Projects, Repository, and ProteinPaint are all applications. Each application handles a specific task such as analysis or
```

visualization and can also be used to refine and build cohorts. Applications are cohort centric and can 
query the GDC API for additional information.

Local and Global filters are available to applications. Local filters are filters that are specific to the application
and are used to refine the data that is displayed in the application. Local filters are those available from the GDC API

> and are typically not the most common. For example in the Mutation Frequency application, the local filters are the gene

Reviewer: qn. i don't quite understand the "not the most common" part. do you mean, e.g. that some cohort filters are waaaay more common, e.g. age?

Author: maybe more frequently/less frequently used?

Reviewer: ah, maybe that would work, though the local filters in Repository are probably highly used. 
or we can avoid overthinking it and just remove mention of whether they are common or frequently used or whatever.

Author: Removing all mention of common/frequency is fine with me.

and mutation type filters. In the figure below the local filters are highlighted in yellow. These filters are used to
refine the input cohort allowing users to drill down to specific genes and mutation types of interest in the cohort.

![Mutation Frequency](./images/developers_guide/mutation_frequency_parts_highlighted.png)

### Local vs Global Filters

The GDC Data Portal application's input can be the current cohort or multiple user defined cohorts. The application then 
allow users to add filters refining the cohort, create new additional cohorts, or display the data in a visualization. 

> Applications  typically have:

Reviewer: ```suggestion
Applications typically have:
```

> * **Local filters**  Refine the data displayed in the application
> * **Global filters**  Defines the cohort
> * **UI Components** Display the data in the application
> * **State** Stores the data displayed in the application
> * **Actions** Updates the state of the application

Reviewer: ```suggestion
* **Local filters**  Refine the data displayed in the application
* **Global filters**  Defines the cohort
* **UI Components** Display the data in the application
* **State** Stores the data displayed in the application
* **Actions** Update the state of the application
```

* Intersection
* Union

The `root` field is a map of filter names (as defined in the GDC API) to filter operation. The filter operation can be
either a single operation or a `FilterSet` object. The `op` field will eventually support either `and` or `or`, however

> at this time only `and` is supported. The `and`operator is used to combine filters using the `and` operator. The `or`

Reviewer: ```suggestion
at this time only `and` is supported. The `and` operator is used to combine filters using the `and` operator. The `or`
```

operator is used to combine filters using the `or` operator. The `FilterSet` object is converted to the appropriate
format for the GDC API when the cohort is saved.

When using the GDC REpresentational State Transfer (REST) API, the FilterSet can be converted into the appropriate 
format using the `filterSetToOperation` function. When using the GDC GraphQL API, the FilterSet can be using the

> `convertFilterSetToGraphQL` function. The API  guide will provide information on what format the filters should be in for the API. Also as the code is in TypeScript,

Reviewer: ```suggestion
`convertFilterSetToGraphQL` function. The API guide will provide information on what format the filters should be in for the API. Also as the code is in TypeScript,
```

page [portal.gdc.cancer.gov/files/0b5a9e7e-8e2e-4b7a-9b7e-ff5d9c5b2b2b](https://portal.gdc.cancer.gov/files/0b5a9e7e-8e2e-4b7a-9b7e-ff5d9c5b2b2b)

```typescript
 const {data: {files} = {}, isFetching} = useGetFilesQuery({
    filters: {
        op: "=",
        content: {
            field: "file_id",
            value: setCurrentFile,
        },
    },
    expand: [
        "cases",
        "cases.annotations",
        "cases.project",
        "cases.samples",
        "cases.samples.portions",
        "cases.samples.portions.analytes",
        "cases.samples.portions.slides",
        "cases.samples.portions.analytes.aliquots",
        "associated_entities",
        "analysis",
        "analysis.input_files",
        "analysis.metadata.read_groups",
        "downstream_analyses",
        "downstream_analyses.output_files",
        "index_files",
    ],
});
```

The `useFileSummary` hook takes several arguments:

* `filters` - The filters to apply to the cases and where the file uuid is passed in
* `expand` - The fields to expand

## Sets: Gene, SSMS, and Case

Sets are supported by the GDC API and are used to create an entity that represents a set of items as a `set_id`. Sets

> are either gene sets, SSM sets, or case sets. All of the GDC APIs support passing sets as a filter parameter.

Reviewer: i wonder if we need to document (later?) the intent and set_type thing for sets.

Author: We should, since that is new, but maybe in another update?

Reviewer: yup - another update

* `useSsmSetCountsQuery`
* `useCaseSetCountsQuery`

## Creating a Cohort

> Depending on the application function, if may be beneficial to create a new cohort. Although the GDC Data Portal SDK provides a
> number of functions for creating a new cohort. It is highly recommended that the application use the provided `Button` and
> `SaveCohortModal` components to create a new cohort. The `Button` and `SaveCohortModal` components are located in

Reviewer: ```suggestion
Depending on the application function, it may be beneficial to create a new cohort. Although the GDC Data Portal SDK provides a
number of functions for creating a new cohort, it is highly recommended that the application use the provided `Button` and
`SaveCohortModal` components to create a new cohort. The `Button` and `SaveCohortModal` components are located in
```

the `@gff/portal-proto` package.

To create a cohort using the SaveCohortModal component the following code can be used:
In summary, the above code flow is:

1. The `ProjectsCohortButton` component renders a button with the label "Save New Cohort"
2. When the button is clicked, it sets the state variable `showSaveCohort` to true, which triggers the rendering of
   the `SaveCohortModal` component.
3. The `SaveCohortModal` component passed:
   * An onClose function that sets the showSaveCohort state variable to false.

>    * A filters prop, which is an object defining the filters for the cohort based on the selected projects.

Reviewer: do we need to surround "filters" with backticks? -- wasn't sure if it's the actual name of the prop

Author: Yes. I will add that.

* `clearCohortFilters`

Note that all of these operations are applied to the current cohort. The current cohort is the cohort that is currently

> being displayed in the Cohort Management Bar. Programmatically current cohort can be accessed via the `selectCurrentCohort` selector.

Reviewer: Maybe... :
"The current cohort can be programmatically accessed..."? (I thought it was ok before the change too)

### Updating, Removing, and Clearing filters

To update the current selected cohort's filter, the `updateCohortFilter` action can be used. The `updateCohortFilter`
action takes two arguments:

```typescript
interface UpdateFilterParams {
    field: string;
    operation: Operation;
}
```

where `field` is the field to update and `operation` is the operation to apply to the field. For example to update the
`cases.project.project_id` field to include the project `TCGA-ACC` the following code can be used:

```typescript
import {useCoreDispatch, updateCohortFilter} from '@gff/core';

const coreDispatch = useCoreDispatch();

coreDispatch(updateCohortFilter({
    field: "cases.project.project_id",
    operation: {
        op: "in",
        content: {
            field: "cases.project.project_id",
            value: ["TCGA-ACC"],
        },
    },
}));
```

This will update the current cohort's filter to include the project `TCGA-ACC`. The `removeCohortFilter` action can be
used to remove a filter from the current cohort. The `removeCohortFilter` action takes a single argument:

```typescript
interface RemoveFilterParams {
    field: string;
}
```

> where `field` is the field to remove. For example to remove the `cases.project.project_id` field from the current
> cohort's filter the following code can be used:

Reviewer: ```suggestion
where `field` is the field to remove. For example, to remove the `cases.project.project_id` field from the current
cohort's filter, the following code can be used:
```

```typescript
import {useCoreDispatch, removeCohortFilter} from '@gff/core';

const coreDispatch = useCoreDispatch();

coreDispatch(removeCohortFilter({
    field: "cases.project.project_id",
}));
```

This will remove the `cases.project.project_id` field from the current cohort's filter. The `clearCohortFilters` action

> can be used to clear all the filters from the current cohort. The `clearCohortFilters` action takes no arguments. For
> example to clear all the filters from the current cohort the following code can be used:

Reviewer: ```suggestion
can be used to clear all the filters from the current cohort. The `clearCohortFilters` action takes no arguments. For
example, to clear all the filters from the current cohort, the following code can be used:
```

This will clear all the filters from the current cohort.

#### Updating the Cohort Name

The cohort name can be updated using the `updateCohortName` action. The `updateCohortName` action takes a single
argument:

```typescript
interface UpdateCohortNameParams {
    name: string;
}
```

> where `name` is the new name for the cohort. For example, to update the current cohort's name to `My Cohort` the
> following code can be used:

Reviewer: ```suggestion
where `name` is the new name for the cohort. For example, to update the current cohort's name to `My Cohort`, the
following code can be used:
```

```typescript
import {useCoreDispatch, updateCohortName} from '@gff/core';

const coreDispatch = useCoreDispatch();

coreDispatch(updateCohortName({
    name: "My Cohort",
}));
```

This will update the current cohort's name to `My Cohort`.

#### Setting the Current Cohort

The current cohort can be set using the `setCurrentCohort` action. The `setCurrentCohort` action takes a single
argument:

```typescript
interface SetCurrentCohortParams {
    cohortId: string;
}
```

> where `cohortId` is the id of the cohort to set as the current cohort. For example to set the cohort with id `1234` as
> the current cohort the following code can be used:

Reviewer: ```suggestion
where `cohortId` is the id of the cohort to set as the current cohort. For example, to set the cohort with id `1234` as
the current cohort, the following code can be used:
```

```typescript
import {useCoreDispatch, setCurrentCohort} from '@gff/core';

const coreDispatch = useCoreDispatch();

coreDispatch(setCurrentCohort({
    cohortId: "1234",
}));
```

This will set the cohort with ID `1234` as the current cohort.

## Total Count Information

Count information can be queried using the `useTotalCounts` hook. This hook takes a number of arguments:

```typescript
import {useTotalCounts} from "@gff/core";

const {data, isFetching, isSuccess, isError} = useTotalCounts();
```

This will return the total counts for the GDC. The data in the response is of the form:

```typescript
interface TotalCounts {
    counts: {
        caseCounts: number;
        fileCounts: number;
        genesCounts: number;
        mutationCounts: number;
        repositoryCaseCounts: number;
        projectsCounts: number;
        primarySiteCounts: number;
    },
    status: DataStatus;
}
```

where `DataStatus` is defined as:

```typescript
export type DataStatus = "uninitialized" | "pending" | "fulfilled" | "rejected";
```

## Application Card Counts

> The application cards show the counts for the data required by it. The data types below are supported:

Reviewer: ```suggestion
The application cards show the counts for the data required by them. The data types below are supported:
```
or...:
"Each application card shows the counts for the data required by it."?

```typescript
import { CountHookRegistry}  from "@gff/core";

...

CountHookRegistry.getInstance().registerHook("ssmCaseCount", useLazySsmsCaseCountQuery);
```

The count function is now registered with it name passed as the first argument to ```registerHook``` , and is used to set the value of the ```countsField``` in the 
application registration described below. An appropriate place to add the registration call is in ```_app.tsx```.

## Component Library

The GDC Data Portal provides a number of components that make it easy to develop applications
for it. These components are located in the `@gff/portal-proto`
package.
In several components, the GDC Data Portal uses the [Mantine](https://mantine.dev/) component library but base components and
encapsulates calls to the GDC API so that developers do not have to.

### Buttons

The GDC Data Portal provides a number of buttons that can be used for various purposes. These buttons are located in
the `@gff/portal-proto` package.
The buttons are:

* `DownloadButton` - A button that can be used to download data from the GDC API.
* `SaveCohortButton` - A button that can be used to save a cohort.

> The `DownloadButton` component is used in the repository application to download data from the GDC API.

Reviewer: ```suggestion
The `DownloadButton` component is used in the Repository application to download data from the GDC API.
```

### Modals

Modals are used to show transitory information or obtain information from the user. The GDC Data Portal provides many
modals that can be used for various purposes. One such modal is the `SaveCohortModal` component mentioned previously.

![img.png](images/developers_guide/components/save_cohort_modal.png)

* `SaveCohortModal` - A modal that can be used to save a cohort.
* Various modals for displaying information on Sets:
  * `CaseSetModal`
  * `GeneSetModal`
  * `MutationSetModal`
* `SaveOrCreateEntityModal` - A modal that can be used to save or create a new entity.

These modals and others, are documented in the Portal 2.0 SDK API documentation.

### Charts

> Basic charts are provided for use within an application, although developers are free to use any  desired charting 

Reviewer: ```suggestion
Basic charts are provided for use within an application, although developers are free to use any desired charting 
```

## Getting Started

The GDC Data Portal 2.0 is a monorepo that contains all the code for the GDC Data Portal. The monorepo is managed
using [lerna](https://lerna.js.org) and [npm](https://www.npmjs.com/), and contains the following packages:

* `@gff/core` - Contains the core components and hooks for the GDC Data Portal.
* `@gff/portal-proto` - Contains the UI components and application framework (using NextJS) for the GDC Data Portal.

Note that in the future, the UI components located in the `@gff/portal-proto` package will be refactored into a 
separate package , and `@gff/portal-proto` will be renamed to `@gff/portal`.

Developers can get started by cloning the repo and following the instructions in
the [README.md](https://github.com/NCI-GDC/gdc-frontend-framework/blob/develop/README.md) file.

## Application Layout

A typical application will have the following layout. The main section of the application is the area where components
like tables, graphs, and other components are displayed. Local filters are displayed on the left side and
depending on the number of facets, will scroll vertically. This is a typical layout but other layouts are possible, like

> in the case of Protein Paint. Applications are encouraged to use vertical space as much as possible, as horizontal

Reviewer: ```suggestion
in the case of ProteinPaint. Applications are encouraged to use vertical space as much as possible, as horizontal
```

        </span>
            </Tooltip>
            {showSaveCohort && (
                <SaveCohortModal
                    onClose={() => setShowSaveCohort(false)}
                    filters={{
                        mode: "and",
                        root: {
                            "cases.project.project_id": {
                                operator: "includes",
                                field: "cases.project.project_id",
                                operands: pickedProjects,
                            },
                        },
                    }}
                />
            )}
        </>
    );
};

export default ProjectsCohortButton;
```

This custom button component uses the state variable `showSaveCohort` to determine if the `SaveCohortModal` component
needs to be shown.
The `SaveCohortModal` component is passed to the current list of projects selected by the user and handles the creation of
the cohort and saving it.

### Application Demo

In addition to the actual application, it can have a demo. The demo can be used to show the
application's functionality and is shown when the demo button is clicked. The demo button is shown when the

> application is registered with `hasDemo: false,` as described in

Reviewer: `hasDemo: true`?

Author: Good catch. Not sure why that was ever written as false.

This file contains an array of registered applications. For example the entry for the Project Center is:

```typescript
import ProjectsIcon from "public/user-flow/icons/crowd-of-users.svg";

...
{
    name: "Projects", 
        icon: (<ProjectsIcon
            width={64}
            height={64}
            viewBox="0 -20 128 128"
            role="img"
            aria-label="Projects icon" />),
        tags: [], 
        hasDemo: false,
        id: "Projects",
        countsField: "caseCount",
        description: "View the Projects available within the GDC and select them for further exploration and analysis.",
}
...
```

The above code registers the Project Center application with the GDC Data Portal. The members of the object are:

* `name` The name of the application
* `icon` The icon as an SVG file, its size and position can be adjusted using the `width`, `height`, and `viewBox`
properties
* `tags` The tags for the application used for searching (which is not currently active)

> * `hasDemo` A boolean indicating if the application has a demo, if so the demo button will be shown

Reviewer: ```suggestion
* `hasDemo` A boolean indicating if the application has a demo. If so, the demo button will be shown.
```

> * `id` The id of the application and needs to match the id of the application registered in

Reviewer: ```suggestion
* `id` The id of the application, which needs to match the id of the application registered in
```

the `createGdcAppWithOwnStore` function
* `countsField` The field to use for the counts in the application, this is used to determine if the application can
be used
* `description` The description of the application
* `noDataTooltip` The tooltip to show if the application has no data

When the application is registered, it will be available in the GDC Data Portal. The application can be accessed by clicking on the
application card.
The visual elements of the card are:

![application_card.png](images/developers_guide/application_card.png)

*Application Card and Associated Elements*

## Source Code Layout

> While Developers have freedom in structuring application code, the following is a recommended layout for an

Reviewer: ```suggestion
While developers have freedom in structuring application code, the following is a recommended layout for an
```

application's source code:

![source code layout](images/developers_guide/app_source_code_layout_fig.png)

*Application Source Code Layout*

# Appendix

## Using Selectors and Hooks

Although a complete guide to React hooks and selectors is out of the scope of this document, a brief
overview of how to use them for application development is provided. For more information on hooks and selectors please see the
[React Hooks](https://react.dev/reference/react/hooks) documentation. As the GDC uses the Redux-toolkit, calls
described in the [Redux Toolkit](https://redux-toolkit.js.org/tutorials/typescript) documentation are used as examples.

### Selectors

Selectors are used to access the state of the GDC Data  Portal's main redux store. Using selectors is the preferred method for
accessing the state of the GDC Data Portal. Selectors are functions that take the state as an argument and return a value.

```typescript
import {useCoreSelector, selectCurrentCohort} from '@gff/core';

const currentCohort = useSelector(selectCurrentCohort);
```

The selector will return the current value of the item in the store. Consult the GDC 2.0 API documentation for a

> complete  list of selectors.

Reviewer: ```suggestion
complete list of selectors.
```

returned is typically a promise that resolves to the data requested. The GDC Data Portal provides a number of hooks for
fetching data from the GDC 2.0 API. These hooks are located in the `@gff/core` package.

```typescript
import {useGeneSymbol} from '@gff/core';

const {data: geneSymbolDict, isSuccess} = useGeneSymbol(
    field === "genes.gene_id" ? facetValues.map((x) => x.toString()) : [],
);
```

GDC Data Portal hooks are designed to work similarly to the RTL Query hooks. The hooks take arguments and return an object,
which contains the data and the status of the query. The status of the query is stored in the `isSuccess` variable.
The data returned from the query is stored in the `data` variable. The object returned from a GDC hook is of the form:

```typescript
{
    data: any;
    isSuccess: boolean;
    isLoading: boolean;
    isError: boolean;
    error: Error;
}
```

where `data` is the data returned from the query, `isSuccess` is a boolean indicating if the query was
successful, `isLoading`
is a boolean indicating if the query is currently loading, `isError` is a boolean indicating if the query resulted in an

> error,  and `error` is the error returned from the query.

Reviewer: ```suggestion
error, and `error` is the error returned from the query.
```
